### PR TITLE
[Merged by Bors] - chore(Data/List): Depend less on big operators

### PR DESCRIPTION
**Pull request by YaelDillies**
* Make `Data.List.Count`, `Data.List.Dedup`, `Data.List.ProdSigma`, `Data.List.Range`, `Data.List.Rotate`, `Data.List.Zip` not depend on `Data.List.BigOperators.Basic`.
* As a consequence, move the big operators lemmas that were in there to `Data.List.BigOperators.Basic`. For the lemmas that were `Nat`-specific, keep a version of them in the original file but stated using `Nat.sum`.
* To help with this, add `Nat.sum_eq_listSum (l : List Nat) : Nat.sum l = l.sum`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/List/ProdSigma.lean [hunk 1]
```diff
@@ -85,11 +85,11 @@ theorem mem_sigma {l₁ : List α} {l₂ : ∀ a, List (σ a)} {a : α} {b : σ
     exists_eq_left, heq_iff_eq, exists_eq_right]
 #align list.mem_sigma List.mem_sigma
 
-theorem length_sigma (l₁ : List α) (l₂ : ∀ a, List (σ a)) :
-    length (l₁.sigma l₂) = (l₁.map fun a => length (l₂ a)).sum := by
+/-- See `List.length_sigma` for the corresponding statement using `List.sum`. -/
```
Comment (eric-wieser) [inline Mathlib/Data/List/ProdSigma.lean:88]: Where is this lemma? I can't find it with Ctrl+F

Comment (YaelDillies) [reply Mathlib/Data/List/ProdSigma.lean:88]: Forgotten, now fixed.

**Review by Vierkantor — COMMENTED** [review]
Once #11729 is merged, could you rebase / merge master? It seems likely to give merge conflicts.

I'm not convinced this move is the best way to go: `List.BigOperators.Basic` is (turning into) a random assortment of lemmas, I definitely have trouble parsing its internal structure. Moreover, feel like `Finset.prod` requires way too many imports already and this PR looks like that amount will increase even more.

On the other hand, the alternative would be lots of files named e.g. `Algebra.BigOperators.List.Dedup` that contain only a few results each is also an unsatisfying outcome. (Especially when we start working with more list operators.) In my opinion, that would be a better option but only marginally. Or do you have a better idea?

**Comment by YaelDillies** [comment]
To me this is a second-order concern: the primary concern is to stop algebra, order theory and finiteness from leaking into each other (this PR is specifically concerned about the algebra → finiteness leak, can you believe we need rings to define finite sets!). Organising each sublibrary is tomorrow's job, not today's.

Concretely, this means:
* These lemmas have to move. They can't stay where they are because they concern algebra but are in finiteness files.
* `Data.List.BigOperators.Basic` is a mess, but already was before, and it's not really this PR's job to make the situation better.

There are avenues to improve `Data.List.BigOperators.Basic` in the future:
* Delete `Data.List.BigOperators.Lemmas` (this is the infamous "too much import for Basic" file every folder has)
* Instead have files graded according to the algebra imports they need. Something like this would be much better than what we have now:
  * Monoid/group lemmas
  * Ring lemmas
  * Funkier stuff
* Move the order stuff to a new `Algebra.Order.BigOperators.List` file (this is already done in #11725)

I could improve the situation in this PR but that would increase the diff.

**Comment by Vierkantor** [comment]
I suppose this PR indeed makes things strictly better than the current situation. Unfortunately my laptop is broken so I can't test that everything ended up moved correctly. @fpvandoorn, would you like to assign yourself and finish the review?

**Comment by fpvandoorn** [comment]
Thanks!

bors merge
bors d+

**Pull request closed without merging**